File: bindings/js/opaque.js
/*
 * Copyright (c) 2021-2023, Alden Torres
 *
 * Licensed under the terms of the MIT license.
 * Copy of the license at https://opensource.org/licenses/MIT
 */

import {
    libecc,
} from "./util.js";

/**
 * Recover the public key related to the input "private_key".
 *
 * @param {Uint8Array} privateKey
 * @return {Uint8Array}
 */
export function opaque_RecoverPublicKey(
    privateKey,
) {

    let publicKey = new Uint8Array(libecc.ecc_opaque_ristretto255_sha512_Npk);

    libecc.ecc_opaque_ristretto255_sha512_RecoverPublicKey(
        publicKey,
        privateKey,
    );

    return publicKey;
}

/**
 * Returns a randomly generated private and public key pair.
 *
 * This is implemented by generating a random "seed", then
 * calling internally DeriveAuthKeyPair.
 *
 * @return object {private_key, public_key}
 */
export function opaque_GenerateAuthKeyPair() {

    let private_key = new Uint8Array(libecc.ecc_opaque_ristretto255_sha512_Nsk);
    let public_key = new Uint8Array(libecc.ecc_opaque_ristretto255_sha512_Npk);

    libecc.ecc_opaque_ristretto255_sha512_GenerateAuthKeyPair(
        private_key,
        public_key,
    );

    return {
        privateKey: private_key,
        publicKey: public_key,
    };
}

/**
 * @param {Uint8Array} password an opaque byte string containing the client's password
 * @param {Uint8Array} blind the OPRF scalar value to use, size:ecc_opaque_ristretto255_sha512_Ns
 * @return object {request, blind}
 */
export function opaque_CreateRegistrationRequestWithBlind(
    password,
    blind,
) {
    let request = new Uint8Array(libecc.ecc_opaque_ristretto255_sha512_REGISTRATIONREQUESTSIZE);

    libecc.ecc_opaque_ristretto255_sha512_CreateRegistrationRequestWithBlind(
        request,
        password, password.length,
        blind,
    );

    return {
        registrationRequest: request,
        blind: blind,
    };
}

/**
 * @param {Uint8Array} password an opaque byte string containing the client's password
 * @return object {request, blind}
 */
export function opaque_CreateRegistrationRequest(
    password
) {
    let request = new Uint8Array(libecc.ecc_opaque_ristretto255_sha512_REGISTRATIONREQUESTSIZE);
    let blind = new Uint8Array(libecc.ecc_opaque_ristretto255_sha512_Ns);

    libecc.ecc_opaque_ristretto255_sha512_CreateRegistrationRequest(
        request,
        blind,
        password, password.length
    );

    return {
        registrationRequest: request,
        blind: blind,
    };
}

/**
 * In order to make this method not to use dynamic memory allocation, there is a
 * limit of credential_identifier to length <= 200.
 *
 * @param {Uint8Array} request a RegistrationRequest structure
 * @param {Uint8Array} server_public_key the server's public key
 * @param {Uint8Array} credential_identifier an identifier that uniquely represents the credential being registered
 * @param {Uint8Array} oprf_seed the server-side seed of Nh bytes used to generate an oprf_key
 * @return {Uint8Array}
 */
export function opaque_CreateRegistrationResponse(
    request,
    server_public_key,
    credential_identifier,
    oprf_seed,
) {
    let response = new Uint8Array(libecc.ecc_opaque_ristretto255_sha512_REGISTRATIONRESPONSESIZE);

    libecc.ecc_opaque_ristretto255_sha512_CreateRegistrationResponse(
        response,
        request,
        server_public_key,
        credential_identifier, credential_identifier.length,
        oprf_seed,
    );

    return response;
}

/**
 * To create the user record used for further authentication, the client
 * executes the following function.
 *
 * @param {Uint8Array} password an opaque byte string containing the client's password
 * @param {Uint8Array} blind the OPRF scalar value used for blinding
 * @param {Uint8Array} response_raw a RegistrationResponse structure
 * @param {Uint8Array} server_identity the optional encoded server identity
 * @param {Uint8Array} client_identity the optional encoded client identity
 * @param {number} mhf the memory hard function to use
 * @param {Uint8Array} mhf_salt the salt to use in the memory hard function computation
 * @param {Uint8Array} nonce size:ecc_opaque_ristretto255_sha512_Nn
 * @return object {record, exportKey}
 */
export function opaque_FinalizeRegistrationRequestWithNonce(
    password,
    blind,
    response_raw,
    server_identity,
    client_identity,
    mhf,
    mhf_salt,
    nonce,
) {
    server_identity = server_identity || new Uint8Array(0);
    client_identity = client_identity || new Uint8Array(0);
    mhf_salt = mhf_salt || new Uint8Array(0);

    let record_raw = new Uint8Array(libecc.ecc_opaque_ristretto255_sha512_REGISTRATIONRECORDSIZE);
    let export_key = new Uint8Array(64);

    libecc.ecc_opaque_ristretto255_sha512_FinalizeRegistrationRequestWithNonce(
        record_raw,
        export_key,
        password, password.length,
        blind,
        response_raw,
        server_identity, server_identity.length,
        client_identity, client_identity.length,
        mhf,
        mhf_salt, mhf_salt.length,
        nonce,
    );

    return {
        registrationRecord: record_raw,
        exportKey: export_key,
    };
}

/**
 * To create the user record used for further authentication, the client
 * executes the following function.
 *
 * @param {Uint8Array} password an opaque byte string containing the client's password
 * @param {Uint8Array} blind the OPRF scalar value used for blinding
 * @param {Uint8Array} response_raw a RegistrationResponse structure
 * @param {Uint8Array} server_identity the optional encoded server identity
 * @param {Uint8Array} client_identity the optional encoded client identity
 * @param {number} mhf the memory hard function to use
 * @param {Uint8Array} mhf_salt the salt to use in the memory hard function computation
 * @return object {record, exportKey}
 */
export function opaque_FinalizeRegistrationRequest(
    password,
    blind,
    response_raw,
    server_identity,
    client_identity,
    mhf,
    mhf_salt,
) {
    server_identity = server_identity || new Uint8Array(0);
    client_identity = client_identity || new Uint8Array(0);
    mhf_salt = mhf_salt || new Uint8Array(0);

    let record_raw = new Uint8Array(libecc.ecc_opaque_ristretto255_sha512_REGISTRATIONRECORDSIZE);
    let export_key = new Uint8Array(64);

    libecc.ecc_opaque_ristretto255_sha512_FinalizeRegistrationRequest(
        record_raw,
        export_key,
        password, password.length,
        blind,
        response_raw,
        server_identity, server_identity.length,
        client_identity, client_identity.length,
        mhf,
        mhf_salt, mhf_salt.length,
    );

    return {
        registrationRecord: record_raw,
        exportKey: export_key,
    };
}

/**
 *
 * @param {Uint8Array} state a ClientState structure
 * @param {Uint8Array} password an opaque byte string containing the client's password
 * @param {Uint8Array} blind
 * @param {Uint8Array} clientNonce
 * @param {Uint8Array} seed
 * @return {Uint8Array} a KE1 message structure
 */
export function opaque_GenerateKE1WithSeed(
    state,
    password,
    blind,
    clientNonce,
    seed,
) {
    let ke1 = new Uint8Array(libecc.ecc_opaque_ristretto255_sha512_KE1SIZE);

    libecc.ecc_opaque_ristretto255_sha512_GenerateKE1WithSeed(
        ke1,
        state,
        password, password.length,
        blind,
        clientNonce,
        seed,
    );

    return ke1;
}

/**
 *
 * @param {Uint8Array} state a ClientState structure
 * @param {Uint8Array} password an opaque byte string containing the client's password
 * @return {Uint8Array} a KE1 message structure
 */
export function opaque_GenerateKE1(
    state,
    password,
) {
    let ke1 = new Uint8Array(libecc.ecc_opaque_ristretto255_sha512_KE1SIZE);

    libecc.ecc_opaque_ristretto255_sha512_GenerateKE1(
        ke1,
        state,
        password, password.length,
    );

    return ke1;
}

/**
 *
 * @param {Uint8Array} state_raw a ClientState structure
 * @param {Uint8Array} client_identity the optional encoded client identity, which is set
 * to client_public_key if not specified
 * @param {Uint8Array} server_identity the optional encoded server identity, which is set
 * to server_public_key if not specified
 * @param {Uint8Array} ke2_raw a KE2 message structure
 * @param {number} mhf
 * @param {Uint8Array} mhf_salt the salt to use in the memory hard function computation
 * @param {Uint8Array} context
 * @return object {ke3, sessionKey, exportKey, finishRet}
 */
export function opaque_GenerateKE3(
    state_raw,
    client_identity,
    server_identity,
    ke2_raw,
    mhf,
    mhf_salt,
    context,
) {
    client_identity = client_identity || new Uint8Array(0);
    server_identity = server_identity || new Uint8Array(0);
    mhf_salt = mhf_salt || new Uint8Array(0);

    let ke3_raw = new Uint8Array(libecc.ecc_opaque_ristretto255_sha512_KE3SIZE);
    let session_key = new Uint8Array(64);
    let export_key = new Uint8Array(64);

    const ret = libecc.ecc_opaque_ristretto255_sha512_GenerateKE3(
        ke3_raw,
        session_key,
        export_key,
        state_raw,
        client_identity, client_identity.length,
        server_identity, server_identity.length,
        ke2_raw,
        mhf,
        mhf_salt, mhf_salt.length,
        context, context.length,
    );

    return {
        ke3: ke3_raw,
        sessionKey: session_key,
        exportKey: export_key,
        result: ret,
    };
}

/**
 *
 * @param {Uint8Array} state_raw a ServerState structure
 * @param {Uint8Array} server_identity the optional encoded server identity, which is set to
 * server_public_key if null
 * @param {Uint8Array} server_private_key the server's private key
 * @param {Uint8Array} server_public_key the server's public key
 * @param {Uint8Array} record_raw the client's RegistrationUpload structure
 * @param {Uint8Array} credential_identifier an identifier that uniquely represents the credential
 * being registered
 * @param {Uint8Array} oprf_seed the server-side seed of Nh bytes used to generate an oprf_key
 * @param {Uint8Array} ke1_raw a KE1 message structure
 * @param {Uint8Array} client_identity
 * @param {Uint8Array} context the application specific context
 * @param {Uint8Array} maskingNonce
 * @param {Uint8Array} serverNonce
 * @param {Uint8Array} seed
 * @return {Uint8Array} a KE2 structure
 */
export function opaque_GenerateKE2WithSeed(
    state_raw,
    server_identity,
    server_private_key,
    server_public_key,
    record_raw,
    credential_identifier,
    oprf_seed,
    ke1_raw,
    client_identity,
    context,
    maskingNonce,
    serverNonce,
    seed,
) {
    server_identity = server_identity || new Uint8Array(0);
    client_identity = client_identity || new Uint8Array(0);
    context = context || new Uint8Array(0);

    let ke2_raw = new Uint8Array(libecc.ecc_opaque_ristretto255_sha512_KE2SIZE);

    libecc.ecc_opaque_ristretto255_sha512_GenerateKE2WithSeed(
        ke2_raw,
        state_raw,
        server_identity, server_identity.length,
        server_private_key,
        server_public_key,
        record_raw,
        credential_identifier, credential_identifier.length,
        oprf_seed,
        ke1_raw,
        client_identity, client_identity.length,
        context, context.length,
        maskingNonce,
        serverNonce,
        seed,
    );

    return ke2_raw;
}

/**
 *
 * @param {Uint8Array} state_raw a ServerState structure
 * @param {Uint8Array} server_identity the optional encoded server identity, which is set to
 * server_public_key if null
 * @param {Uint8Array} server_private_key the server's private key
 * @param {Uint8Array} server_public_key the server's public key
 * @param {Uint8Array} record_raw the client's RegistrationUpload structure
 * @param {Uint8Array} credential_identifier an identifier that uniquely represents the credential
 * being registered
 * @param {Uint8Array} oprf_seed the server-side seed of Nh bytes used to generate an oprf_key
 * @param {Uint8Array} ke1_raw a KE1 message structure
 * @param {Uint8Array} client_identity
 * @param {Uint8Array} context the application specific context
 * @return {Uint8Array} a KE2 structure
 */
export function opaque_GenerateKE2(
    state_raw,
    server_identity,
    server_private_key,
    server_public_key,
    record_raw,
    credential_identifier,
    oprf_seed,
    ke1_raw,
    client_identity,
    context,
) {
    server_identity = server_identity || new Uint8Array(0);
    client_identity = client_identity || new Uint8Array(0);
    context = context || new Uint8Array(0);

    let ke2_raw = new Uint8Array(libecc.ecc_opaque_ristretto255_sha512_KE2SIZE);

    libecc.ecc_opaque_ristretto255_sha512_GenerateKE2(
        ke2_raw,
        state_raw,
        server_identity, server_identity.length,
        server_private_key,
        server_public_key,
        record_raw,
        credential_identifier, credential_identifier.length,
        oprf_seed,
        ke1_raw,
        client_identity, client_identity.length,
        context, context.length,
    );

    return ke2_raw;
}

/**
 * @param {Uint8Array} state_raw a ServerState structure
 * @param {Uint8Array} ke3_raw a KE3 structure
 * @return object {session_key, finish_ret}
 */
export function opaque_ServerFinish(
    state_raw,
    ke3_raw
) {
    let session_key = new Uint8Array(64);

    const ret = libecc.ecc_opaque_ristretto255_sha512_ServerFinish(
        session_key,
        state_raw,
        ke3_raw,
    );

    return {
        sessionKey: session_key,
        result: ret,
    };
}

File: bindings/jvm/src/main/java/org/ssohub/crypto/ecc/opaque/Opaque.java
/*
 * Copyright (c) 2023, Alden Torres
 *
 * Licensed under the terms of the MIT license.
 * Copy of the license at https://opensource.org/licenses/MIT
 */

package org.ssohub.crypto.ecc.opaque;

import org.ssohub.crypto.ecc.Data;
import org.ssohub.crypto.ecc.ristretto255.R255Scalar;

import static org.ssohub.crypto.ecc.libecc.*;

/**
 * @author aldenml
 */
public final class Opaque {

    private Opaque() {
    }

    public static GenerateAuthKeyPairResult generateAuthKeyPair() {

        byte[] privateKey = new byte[ecc_opaque_ristretto255_sha512_Nsk];
        byte[] publicKey = new byte[ecc_opaque_ristretto255_sha512_Npk];

        ecc_opaque_ristretto255_sha512_GenerateAuthKeyPair(
            privateKey,
            publicKey
        );

        return new GenerateAuthKeyPairResult(
            new OpaqueSk(new Data(privateKey)),
            new OpaquePk(new Data(publicKey))
        );
    }

    public static OpaquePk recoverPublicKey(OpaqueSk privateKey) {

        byte[] publicKey = new byte[ecc_opaque_ristretto255_sha512_Npk];

        ecc_opaque_ristretto255_sha512_RecoverPublicKey(
            publicKey,
            privateKey.toBytes()
        );

        return new OpaquePk(new Data(publicKey));
    }

    public static CreateRegistrationRequestResult createRegistrationRequestWithBlind(
        Data password,
        R255Scalar blind
    ) {
        byte[] passwordBytes = password.toBytes();

        byte[] request = new byte[ecc_opaque_ristretto255_sha512_REGISTRATIONREQUESTSIZE];

        ecc_opaque_ristretto255_sha512_CreateRegistrationRequestWithBlind(
            request,
            passwordBytes,
            passwordBytes.length,
            blind.toBytes()
        );

        return new CreateRegistrationRequestResult(
            new RegistrationRequest(new Data(request)),
            new R255Scalar(blind.getData())
        );
    }

    public static CreateRegistrationRequestResult createRegistrationRequest(
        Data password
    ) {
        byte[] passwordBytes = password.toBytes();

        byte[] request = new byte[ecc_opaque_ristretto255_sha512_REGISTRATIONREQUESTSIZE];
        byte[] blind = new byte[ecc_opaque_ristretto255_sha512_Ns];

        ecc_opaque_ristretto255_sha512_CreateRegistrationRequest(
            request,
            blind,
            passwordBytes,
            passwordBytes.length
        );

        return new CreateRegistrationRequestResult(
            new RegistrationRequest(new Data(request)),
            new R255Scalar(new Data(blind))
        );
    }

    public static RegistrationResponse createRegistrationResponse(
        RegistrationRequest registrationRequest,
        OpaquePk serverPublicKey,
        Data credentialIdentifier,
        Data oprfSeed
    ) {
        if (oprfSeed.size() != ecc_opaque_ristretto255_sha512_Nh)
            throw new IllegalArgumentException("oprf seed should be of size: " + ecc_opaque_ristretto255_sha512_Nh);

        byte[] credentialIdentifierBytes = credentialIdentifier.toBytes();

        byte[] response = new byte[ecc_opaque_ristretto255_sha512_REGISTRATIONRESPONSESIZE];

        ecc_opaque_ristretto255_sha512_CreateRegistrationResponse(
            response,
            registrationRequest.toBytes(),
            serverPublicKey.toBytes(),
            credentialIdentifierBytes,
            credentialIdentifierBytes.length,
            oprfSeed.toBytes()
        );

        return new RegistrationResponse(new Data(response));
    }

    public static FinalizeRegistrationRequestResult finalizeRegistrationRequestWithNonce(
        Data password,
        R255Scalar blind,
        RegistrationResponse registrationResponse,
        Data serverIdentity,
        Data clientIdentity,
        MHF mhf,
        OpaqueSeed nonce
    ) {
        byte[] passwordBytes = password.toBytes();
        byte[] serverIdentityBytes = serverIdentity != null ? serverIdentity.toBytes() : new byte[0];
        byte[] clientIdentityBytes = clientIdentity != null ? clientIdentity.toBytes() : new byte[0];

        byte[] record = new byte[ecc_opaque_ristretto255_sha512_REGISTRATIONRECORDSIZE];
        byte[] exportKey = new byte[ecc_opaque_ristretto255_sha512_Nh];

        ecc_opaque_ristretto255_sha512_FinalizeRegistrationRequestWithNonce(
            record,
            exportKey,
            passwordBytes,
            passwordBytes.length,
            blind.toBytes(),
            registrationResponse.toBytes(),
            serverIdentityBytes,
            serverIdentityBytes.length,
            clientIdentityBytes,
            clientIdentityBytes.length,
            mhf.intValue,
            nonce.toBytes()
        );

        return new FinalizeRegistrationRequestResult(
            new RegistrationRecord(new Data(record)),
            new Data(exportKey)
        );
    }

    public static FinalizeRegistrationRequestResult finalizeRegistrationRequest(
        Data password,
        R255Scalar blind,
        RegistrationResponse registrationResponse,
        Data serverIdentity,
        Data clientIdentity,
        MHF mhf
    ) {
        byte[] passwordBytes = password.toBytes();
        byte[] serverIdentityBytes = serverIdentity != null ? serverIdentity.toBytes() : new byte[0];
        byte[] clientIdentityBytes = clientIdentity != null ? clientIdentity.toBytes() : new byte[0];

        byte[] record = new byte[ecc_opaque_ristretto255_sha512_REGISTRATIONRECORDSIZE];
        byte[] exportKey = new byte[ecc_opaque_ristretto255_sha512_Nh];

        ecc_opaque_ristretto255_sha512_FinalizeRegistrationRequest(
            record,
            exportKey,
            passwordBytes,
            passwordBytes.length,
            blind.toBytes(),
            registrationResponse.toBytes(),
            serverIdentityBytes,
            serverIdentityBytes.length,
            clientIdentityBytes,
            clientIdentityBytes.length,
            mhf.intValue
        );

        return new FinalizeRegistrationRequestResult(
            new RegistrationRecord(new Data(record)),
            new Data(exportKey)
        );
    }

    public static KE1 clientInitWithSecrets(
        ClientState state,
        Data password,
        R255Scalar blind,
        OpaqueSeed clientNonce,
        OpaqueSk clientSecret,
        OpaquePk clientKeyshare
    ) {
        byte[] passwordBytes = password.toBytes();

        byte[] ke1 = new byte[ecc_opaque_ristretto255_sha512_KE1SIZE];

        ecc_opaque_ristretto255_sha512_3DH_ClientInitWithSecrets(
            ke1,
            state.data(),
            passwordBytes,
            passwordBytes.length,
            blind.toBytes(),
            clientNonce.toBytes(),
            clientSecret.toBytes(),
            clientKeyshare.toBytes()
        );

        return new KE1(new Data(ke1));
    }

    public static KE1 clientInit(
        ClientState state,
        Data password
    ) {
        byte[] passwordBytes = password.toBytes();

        byte[] ke1 = new byte[ecc_opaque_ristretto255_sha512_KE1SIZE];

        ecc_opaque_ristretto255_sha512_3DH_ClientInit(
            ke1,
            state.data(),
            passwordBytes,
            passwordBytes.length
        );

        return new KE1(new Data(ke1));
    }

    public static KE2 serverInitWithSecrets(
        ServerState state,
        Data serverIdentity,
        OpaqueSk serverPrivateKey,
        OpaquePk serverPublicKey,
        RegistrationRecord record,
        Data credentialIdentifier,
        Data oprfSeed,
        KE1 ke1,
        Data clientIdentity,
        Data context,
        OpaqueSeed maskingNonce,
        OpaqueSeed serverNonce,
        OpaqueSk serverSecret,
        OpaquePk serverKeyshare
    ) {
        if (oprfSeed.size() != ecc_opaque_ristretto255_sha512_Nh)
            throw new IllegalArgumentException("oprf seed should be of size: " + ecc_opaque_ristretto255_sha512_Nh);

        byte[] serverIdentityBytes = serverIdentity != null ? serverIdentity.toBytes() : new byte[0];
        byte[] clientIdentityBytes = clientIdentity != null ? clientIdentity.toBytes() : new byte[0];
        byte[] credentialIdentifierBytes = credentialIdentifier.toBytes();
        byte[] contextBytes = context.toBytes();

        byte[] ke2 = new byte[ecc_opaque_ristretto255_sha512_KE2SIZE];

        ecc_opaque_ristretto255_sha512_3DH_ServerInitWithSecrets(
            ke2,
            state.data(),
            serverIdentityBytes,
            serverIdentityBytes.length,
            serverPrivateKey.toBytes(),
            serverPublicKey.toBytes(),
            clientIdentityBytes,
            clientIdentityBytes.length,
            record.toBytes(),
            credentialIdentifierBytes,
            credentialIdentifierBytes.length,
            oprfSeed.toBytes(),
            ke1.toBytes(),
            contextBytes,
            contextBytes.length,
            maskingNonce.toBytes(),
            serverNonce.toBytes(),
            serverSecret.toBytes(),
            serverKeyshare.toBytes()
        );

        return new KE2(new Data(ke2));
    }

    public static KE2 serverInit(
        ServerState state,
        Data serverIdentity,
        OpaqueSk serverPrivateKey,
        OpaquePk serverPublicKey,
        RegistrationRecord record,
        Data credentialIdentifier,
        Data oprfSeed,
        KE1 ke1,
        Data clientIdentity,
        Data context
    ) {
        if (oprfSeed.size() != ecc_opaque_ristretto255_sha512_Nh)
            throw new IllegalArgumentException("oprf seed should be of size: " + ecc_opaque_ristretto255_sha512_Nh);

        byte[] serverIdentityBytes = serverIdentity != null ? serverIdentity.toBytes() : new byte[0];
        byte[] clientIdentityBytes = clientIdentity != null ? clientIdentity.toBytes() : new byte[0];
        byte[] credentialIdentifierBytes = credentialIdentifier.toBytes();
        byte[] contextBytes = context.toBytes();

        byte[] ke2 = new byte[ecc_opaque_ristretto255_sha512_KE2SIZE];

        ecc_opaque_ristretto255_sha512_3DH_ServerInit(
            ke2,
            state.data(),
            serverIdentityBytes,
            serverIdentityBytes.length,
            serverPrivateKey.toBytes(),
            serverPublicKey.toBytes(),
            clientIdentityBytes,
            clientIdentityBytes.length,
            record.toBytes(),
            credentialIdentifierBytes,
            credentialIdentifierBytes.length,
            oprfSeed.toBytes(),
            ke1.toBytes(),
            contextBytes,
            contextBytes.length
        );

        return new KE2(new Data(ke2));
    }

    public static ClientFinishResult clientFinish(
        ClientState state,
        Data clientIdentity,
        Data serverIdentity,
        KE2 ke2,
        MHF mhf,
        Data context
    ) {
        byte[] serverIdentityBytes = serverIdentity != null ? serverIdentity.toBytes() : new byte[0];
        byte[] clientIdentityBytes = clientIdentity != null ? clientIdentity.toBytes() : new byte[0];
        byte[] contextBytes = context.toBytes();

        byte[] ke3 = new byte[ecc_opaque_ristretto255_sha512_KE3SIZE];
        byte[] sessionKey = new byte[64];
        byte[] exportKey = new byte[64];

        int result = ecc_opaque_ristretto255_sha512_3DH_ClientFinish(
            ke3,
            sessionKey,
            exportKey,
            state.data(),
            clientIdentityBytes,
            clientIdentityBytes.length,
            serverIdentityBytes,
            serverIdentityBytes.length,
            ke2.toBytes(),
            mhf.intValue,
            contextBytes,
            contextBytes.length
        );

        return new ClientFinishResult(
            new KE3(new Data(ke3)),
            new Data(sessionKey),
            new Data(exportKey),
            result
        );
    }

    public static ServerFinishResult serverFinish(
        ServerState state,
        KE3 ke3
    ) {
        byte[] sessionKey = new byte[64];

        int result = ecc_opaque_ristretto255_sha512_3DH_ServerFinish(
            sessionKey,
            state.data(),
            ke3.toBytes()
        );

        return new ServerFinishResult(
            new Data(sessionKey),
            result
        );
    }

    public enum MHF {
        IDENTITY(0),
        SCRYPT(1);

        final int intValue;

        MHF(int intValue) {
            this.intValue = intValue;
        }
    }
}
